frontend/src/main/scala/bloop/engine/tasks/Task.scala

```diff
@@ -3,16 +3,17 @@ package bloop.engine.tasks
 import scala.collection.mutable
 import scala.concurrent.Await
 import scala.concurrent.duration.Duration
-import scala.util.control.NonFatal
+
+import scala.util.{Failure, Success, Try}
 
 import monix.eval.{Task => MTask}
 import monix.execution.{CancelableFuture, Scheduler}
 
 /**
  * An asynchronous computation, possibly depending on other computations.
  */
-class Task[+T] private (private val underlying: MTask[Result[T]],
-                        origTask: MTask[Result[T]],
+class Task[+T] private (private val underlying: MTask[Try[T]],
```

Comment:
Something feels wrong about this data structure. This seems to be closer to a tree rather than `Task`

```scala
case class Tree[T](node: Task[T], children: Seq[Tree[_]])
```

`children` might even be ListBuffer. `gatherUnordered` can be run at the end when you want to build your final `Task`.

The methods `map`/`flatMap` are not behaving the way I would expect them to, because it only applies to the underlying task and not dependencies.

Reply:
I recommend using `Dag[Task[T]` here, `Dag` was created for this purpose and I find it a very good structure to deal with.

Reply:
@olafurpg Thanks, I've changed the tasks to call `gatherUnordered` at the end when building the final task as you've suggested. It's indeed clearer, and it fixed another issue.

I don't agree about the dependencies being a mutable list or about `map` and `flatMap`. I don't think that having the dependencies mutable solves any problem, but it makes it harder to reason about what happens in this case:

```scala
val t0 = Task { 0 }
val t1 = Task { 1 }
val t2 = t1.map(_ + 1)
t1.dependsOn(t0)
t2.runAsync // Should that trigger t0?
```

I also disagree with your comment about `map` and `flatMap`. The dependencies are arbitrary tasks, I don't see how we could apply an operation on all of them? It's also how these operations are defined on sbt's tasks, for instance.

Reply:
@jvican The problem with `Dag[T]` is that all its elements have type `T`, it's not the case here.